chain/subscriber.go

```diff
@@ -0,0 +1,499 @@
+			continue // ignore old announcements
+		}
+		chain.ForEachHostAnnouncement(b, func(hk types.PublicKey, ha chain.HostAnnouncement) {
+			if err != nil {
+				return // error occurred
+			}
+			if ha.NetAddress == "" {
+				return // ignore
+			}
+			err = tx.UpdateHost(hk, ha, cau.State.Index.Height, b.ID(), b.Timestamp)
+		})
+		if err != nil {
+			return fmt.Errorf("failed to update host: %w", err)
+		}
+
+		// v1 contracts
+		cau.ForEachFileContractElement(func(fce types.FileContractElement, rev *types.FileContractElement, resolved, valid bool) {
+			if err != nil {
+				return // error occurred
+			}
+			curr := &revision{
+				revisionNumber: fce.FileContract.RevisionNumber,
+				fileSize:       fce.FileContract.Filesize,
+			}
+			if rev != nil {
+				curr.revisionNumber = rev.FileContract.RevisionNumber
+				curr.fileSize = rev.FileContract.Filesize
+			}
+			err = s.updateContract(tx, cau.State.Index, types.FileContractID(fce.ID), nil, curr, resolved, valid)
+		})
+		if err != nil {
+			return fmt.Errorf("failed to process v1 contracts: %w", err)
+		}
+
+		// v2 contracts
+		cau.ForEachV2FileContractElement(func(fce types.V2FileContractElement, rev *types.V2FileContractElement, res types.V2FileContractResolutionType) {
+			if err != nil {
+				return // error occurred
+			}
+			curr := &revision{
+				revisionNumber: fce.V2FileContract.RevisionNumber,
+				fileSize:       fce.V2FileContract.Filesize,
+			}
+			if rev != nil {
+				curr.revisionNumber = rev.V2FileContract.RevisionNumber
+				curr.fileSize = rev.V2FileContract.Filesize
+			}
+			resolved, valid := checkFileContract(fce, res)
+			err = s.updateContract(tx, cau.State.Index, types.FileContractID(fce.ID), nil, curr, resolved, valid)
+		})
+		if err != nil {
+			return fmt.Errorf("failed to process v1 contracts: %w", err)
+		}
+	}
+	return
+}
+
+func (s *Subscriber) isKnownContract(fcid types.FileContractID) bool {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	_, known := s.knownContracts[fcid]
+	return known
+}
+
+func (s *Subscriber) revertChainUpdate(tx ChainUpdateTx, cru chain.RevertUpdate) (err error) {
+	// v1 contracts
+	cru.ForEachFileContractElement(func(fce types.FileContractElement, rev *types.FileContractElement, resolved, valid bool) {
+		if err != nil {
+			return // error occurred
+		}
+
+		var prev, curr *revision
+		if rev != nil {
+			curr = &revision{
+				revisionNumber: rev.FileContract.RevisionNumber,
+				fileSize:       rev.FileContract.Filesize,
+			}
+		}
+		prev = &revision{
+			revisionNumber: fce.FileContract.RevisionNumber,
+			fileSize:       fce.FileContract.Filesize,
+		}
+		err = s.updateContract(tx, cru.State.Index, types.FileContractID(fce.ID), prev, curr, resolved, valid)
+	})
+	if err != nil {
+		return fmt.Errorf("failed to revert v1 contract: %w", err)
+	}
+
+	// v2 contracts
+	cru.ForEachV2FileContractElement(func(fce types.V2FileContractElement, rev *types.V2FileContractElement, res types.V2FileContractResolutionType) {
```

Comment:
Was there a reason for these `ForEach` types not to take functions that return an error? I can see why if you foreach state into some kind of state keeping type, like we used to do, but now in practice it's kind of annoying these `ForEach` functions don't handle errors differently.